[BOLT][AArch64] Introduce SPE mode in BasicAggregation

paschalis-mpeis opened the pull request:
BOLT gains the ability to process Arm SPE data using the `BasicAggregation` format.

Example usage is:
```bash
perf2bolt -p perf.data -o perf.boltdata --nl --spe BINARY
```

New branch data and compatibility:
---
`perf` since Linux 6.13 reports for SPE branch pairs (`PC` → `TGT`) where:
- `PC`:
  + it is the source branch; may be taken or not-taken.
  + Due to the nature of how SPE operates and what it can collect, any filtering on the
     `PC` (i.e.,  to consider only the taken branches) would result in a data loss that
       BOLT cannot later infer.
- `TGT`:
  + it is the target address of the destination block.
  + this is the new information that `perf` can now report.

DataAggregator processes this information by creating two basic samples.
Any other event types will have `ADDR` field set to `0x0`. For those a single sample
will be created.
Such events can be either SPE or non-SPE, like `l1d-access` and `cycles` respectively.

The format of the input perf entries is:
```
PID   EVENT-TYPE   ADDR   IP
```

When on SPE mode and:
- host is not `AArch64`, BOLT will exit with a relevant message
- `ADDR` field is unavailable, BOLT will exit with a relevant message
- no branch pairs were recorded, BOLT will present a warning

Examples of generating profiling data for the SPE mode:
---
Profiles can be captured with perf on AArch64 machines with SPE enabled.
They can be combined with other events, SPE or not.
In the future we might restrict processing to just the branch packets.

Capture only SPE branch data events:
```bash
perf record -e 'arm_spe_0/branch_filter=1/u' -- BINARY
```

Using more filters, some jitter, and specify count to control overheads/quality:
```bash
perf record -e 'arm_spe_0/branch_filter=1,load_filter=0,store_filter=0,jitter=1/u' -c 10007 -- BINARY
```

Capture any SPE events:
```bash
perf record -e 'arm_spe_0//u' -- BINARY
```

Capture any SPE events and cycles
```bash
perf record -e 'arm_spe_0//u' -e cycles:u -- BINARY
```

paschalis-mpeis committed:
[BOLT][AArch64] Introduce SPE mode in BasicAggregation

BOLT gains the ability to process branch target information generated by
Arm SPE data, using the `BasicAggregation` format.

Example usage is:
```bash
perf2bolt -p perf.data -o perf.boltdata --nl --spe BINARY
```

New branch data and compatibility:
---
SPE branch entries in perf data contain a branch pair (`IP` -> `ADDR`)
for the source and destination branches. DataAggregator processes those
by creating two basic samples. Any other event types will have `ADDR`
field set to `0x0`. For those a single sample will be created. Such
events can be either SPE or non-SPE, like `l1d-access` and `cycles`
respectively.

The format of the input perf entries is:
```
PID   EVENT-TYPE   ADDR   IP
```

When on SPE mode and:
- host is not `AArch64`, BOLT will exit with a relevant message
- `ADDR` field is unavailable, BOLT will exit with a relevant message
- no branch pairs were recorded, BOLT will present a warning

Examples of generating profiling data for the SPE mode:
---
Profiles can be captured with perf on AArch64 machines with SPE enabled.
They can be combined with other events, SPE or not.

Capture only SPE branch data events:
```bash
perf record -e 'arm_spe_0/branch_filter=1/u' -- BINARY
```

Capture any SPE events:
```bash
perf record -e 'arm_spe_0//u' -- BINARY
```

Capture any SPE events and cycles
```bash
perf record -e 'arm_spe_0//u' -e cycles:u -- BINARY
```

More filters, jitter, and specify count to control overheads/quality.
```bash
perf record -e 'arm_spe_0/branch_filter=1,load_filter=0,store_filter=0,jitter=1/u' -c 10007 -- BINARY
```

paschalis-mpeis committed:
clang-format fix

yota9 commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -1226,6 +1247,66 @@ ErrorOr<DataAggregator::PerfBasicSample> DataAggregator::parseBasicSample() {
+    return std::make_pair(EmptySample, EmptySample);
+  }
+
+  while (checkAndConsumeFS()) {
```
```suggestion
  while (checkAndConsumeFS());
```

paschalis-mpeis replied:
Thanks for the suggestion. Could do that but clang-format would require the semicolon to be on the next line anyway. There is similar code to this in other parsing functions like [parseBasicSample()](https://github.com/llvm/llvm-project/blob/61032dec95da15bb7ff646a0178e08f2bb9347e4/bolt/lib/Profile/DataAggregator.cpp#L1211).

yota9 commented on bolt/tools/driver/llvm-bolt.cpp:
```diff
@@ -51,6 +51,8 @@ static cl::opt<std::string> InputFilename(cl::Positional,
                                           cl::Required, cl::cat(BoltCategory),
                                           cl::sub(cl::SubCommand::getAll()));
 
+extern cl::opt<bool> ArmSPE;
```
Probably should be moved to CommandLineOpts.h

yota9 commented on bolt/tools/driver/llvm-bolt.cpp:
```diff
@@ -245,6 +247,13 @@ int main(int argc, char **argv) {
       RewriteInstance &RI = *RIOrErr.get();
+
+      if (opts::AggregateOnly && !RI.getBinaryContext().isAArch64() &&
+          opts::ArmSPE == 1) {
```
```suggestion
          opts::ArmSPE) {
```

paschalis-mpeis commented:
This PR is an implementation of the (4a) approach of:
- #115333

We did some limited, quick testing and there was no clear winner between the two approaches, but the `--spe` flag is introduced in a way to accommodate both.

I believe @kaadam  had some work on (4b)? Maybe at some point we could additionally have that merged, and community can test on a wider set of apps/workloads. I believe there won't be dramatic performance changes.

Please give SPE a try along with this patch and report any feedback. To check if SPE is available on your machine, see point [(3)](https://github.com/llvm/llvm-project/issues/115333) on the issue. Let us know if more information is needed on how to enable or use SPE!

yota9 commented on bolt/test/perf2bolt/AArch64/perf2bolt-spe.test:
```diff
@@ -0,0 +1,14 @@
+
+CHECK-SPE-NO-LBR: PERF2BOLT: Starting data aggregation job
+
+RUN: perf record -e cycles -q -o %t.perf.data -- %t.exe
```
Maybe it would be better to add static perf record rather then launching perf every time for this test...

paschalis-mpeis replied:
Agreed, I would have preferred that as well. I was not able to work around this for the `CHECK-SPE-LBR` test.

yota9 commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -1226,6 +1247,66 @@ ErrorOr<DataAggregator::PerfBasicSample> DataAggregator::parseBasicSample() {
+  ErrorOr<uint64_t> AddrResTo = parseHexField(FieldSeparator);
+  if (std::error_code EC = AddrResTo.getError())
+    return EC;
+  consumeAllRemainingFS();
```
Please add spaces after  coditions, loops with or without  {};

yota9 reviewed (commented):
Thanks for your amazing job!

paschalis-mpeis commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -1226,6 +1247,66 @@ ErrorOr<DataAggregator::PerfBasicSample> DataAggregator::parseBasicSample() {
+
+  // Show more meaningful event names on boltdata.
+  if (Event->str() == "instructions:")
+    Event = *AddrResTo != 0x0 ? "branch-spe:" : "instruction-spe:";
```
Just adding that this might cause some incompatibility with [instruction count normalization](https://github.com/llvm/llvm-project/blob/5d38a3406b11c70e6f0d1a880b78ed404aba2c36/bolt/lib/Profile/DataReader.cpp#L566) detection. Unrelated to this patch, but I see that `cycles:u` are not detected by that pattern either.

---

An alternative approach to 'tag' those events could be to avoid using `--itrace=i1i` for aggregation in perf2bolt, keep the event name given by `perf` and create the additional sample only for the `branch` type.

paschalis-mpeis committed:
Addressing reviewers (1)

paschalis-mpeis commented:
Hey @yota9,

Thanks a lot for your review!

I addressed your comments except [this one](https://github.com/llvm/llvm-project/pull/120741#discussion_r1894153391) (left a comment there).
Please have another look and let me know of any further changes.

maksfb commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -180,22 +187,29 @@ void DataAggregator::start() {
+      exit(1);
+    }
+    launchPerfProcess("branch events with SPE", MainEventsPPI,
+                      "script -F pid,event,ip,addr --itrace=i1i",
```
What's the expected output of this perf script? On my system `ip` is always 0.

maksfb replied:
The zeros on my side could be b/c we don't have kernel patches listed in #115333.

paschalis-mpeis replied:
Hey Maks,

We expect:
- branch entries to have non-zero values for both `addr` and `ip`
- non-branch entries to have non-zero value only for `ip`

This [unit test](https://github.com/llvm/llvm-project/pull/120741/files#diff-b16a95c2248de4f899b4bea1591c813c69caed2f5e3702ea53e0a80d75b1a79cR133) illustrates such an output.


> On my system ip is always 0 for the data collected  ...

The zeros you are seeing actually refer to `addr` and not `ip`. It is a bit confusing but the printed order is different than the fields order specified in CLI. Can verify using:
```bash
perf script -F pid,event,ip ..
perf script -F pid,event,addr .. # seeing zeros on unpatched perf
```

> The zeros on my side could be b/c we don't have kernel patches listed in [#115333](https://github.com/llvm/llvm-project/issues/115333).

Correct. We believe these patches might make it to the next kernel RC.  
In the meantime, I've added some [sample instructions](https://github.com/llvm/llvm-project/issues/115333#issuecomment-2597809642) on the issue to quickly use it locally, w/o requiring to re-flush a kernel.

paschalis-mpeis replied:
Update: As of yesterday, the required patches are part of **Linux 6.13** [[kernel diff](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/diff/?id=v6.13&id2=v6.12&dt=2)].

aaupov commented:
Hi Paschalis, thank you for working on this.
The benefit that SPE has over IP sampling is the edge frequency information. So instead of creating two basic (IP) samples we should create branch samples (LBR) with stack depth one. Branch samples are later attached to CFG edges. This should improve the resulting performance when using SPE profiling.

aaupov commented:
> This PR is an implementation of the (4a) approach of:
> 
> * [[AArch64] BOLT does not support SPE branch data #115333](https://github.com/llvm/llvm-project/issues/115333)
> 
> We did some limited, quick testing and there was no clear winner between the two approaches, but the `--spe` flag is introduced in a way to accommodate both.

Missed this comment. Am I reading it right that you didn't see a perf difference between registering SPE as two basic events or one branch event? In this case, can you please try `-infer-fall-throughs` option with the latter?

maksfb commented on bolt/tools/driver/llvm-bolt.cpp:
```diff
@@ -245,6 +247,13 @@ int main(int argc, char **argv) {
+
+      if (opts::AggregateOnly && !RI.getBinaryContext().isAArch64() &&
+          opts::ArmSPE == 1) {
+        errs() << "BOLT-ERROR: -spe is available only on AArch64.\n";
```
nit: use `ToolName` instead of `"BOLT-ERROR"`.

paschalis-mpeis committed:
Addressing reviewers (2)

paschalis-mpeis commented:
Hey Amir and Maks,

Thank you for taking a look at this!

> Am I reading it right that you didn't see a perf difference between registering SPE as two basic events or one branch event? In this case, can you please try -infer-fall-throughs option with the latter?

Correct, in some preliminary internal tests we found both approaches to be close to each other.

Thanks for your suggestion to use `-infer-fall-throughs`. I thought LBR mode was inferring fall-through branches by default. But it looks like this has to be manually specified?

**Let me share my understanding on the LBR format to see if I got this right:**
Each LBR event gets a contiguous stack of taken branches. And any other branches that may lay in between them are **known** to be fall-throughs, which BOLT can infer.  eg, if we have:
- $\bf\textsf{\color{blue}TK1}$ -> $\textsf{\color{blue}TK2}$ -> $\textsf{\color{blue}TK3}$, then BOLT can propagate CFG hotness to:
- $\textsf{\color{blue}TK1}$ -> FT1a -> FT1b .. -> $\textsf{\color{blue}TK2}$ -> FT2a -> FT2b .. -> $\textsf{\color{blue}TK3}$

SPE on the other hand is a statistical sampling method, meaning all collected packets are not captured contiguously. Each pair comes from a packet that looks like:

```
.  00000040:         PC 0xABC el2 ns=1
.  00000049:         PAD
.  00000053:         B COND
.  00000055:         EV RETIRED NOT-TAKEN
.  0000005a:         LAT 7 ISSUE
.  0000005d:         LAT 8 TOT
.  00000060:         TGT 0xDEF el2 ns=1
.  00000069:         PAD
.  00000077:         TS 1234
```
(note: you can inspect native SPE packets w/ `perf script -D`)

From this example we have  `0xABC` -> `0xDEF` (a src/tgt pair), where `0xABC` is a branch that was NOT-TAKEN. 
The tgt `0xDEF` is a target address of some block (ie, not a branch). We have no information whether the branch of that target block will be taken or not. Therefore, my understanding is that we cannot infer any branches in-between src/tgt. And I believe that is why we found the two approaches to be close to each other.

Please do share your thoughts on this.

Do you think there are any other benefits when using the LBR format? It can additionally utilize prediction information (miss/hit), but we haven't found this to be that beneficial for the quite-limited SPE branch data (when compared to LBR traces).

aaupov commented:
@paschalis-mpeis is there a way to configure SPE to only collect taken branches? My impression was that it's possible, e.g. based on this: https://community.arm.com/arm-community-blogs/b/tools-software-ides-blog/posts/introduction-to-statistical-profiling-support-in-streamline
> Event packets, which provide important information about each sampled instruction.
This information includes:
...
Was a mis-predicted or not-taken branch

But I couldn't find any info regarding configuring perf filter to collect it.

aaupov commented:
> Let me share my understanding on the LBR format to see if I got this right:
Each LBR event gets a contiguous stack of taken branches. And any other branches that may lay in between them are known to be fall-throughs, which BOLT can infer. eg, if we have:

Right, with taken branch stacks, we automatically "infer" fall-throughs between entries. But if we were to collect single branch records, as I assume should be possible with SPE, we'd need to use `infer-fall-throughs` to perform inference on a CFG _after_ edge counters are attached.

paschalis-mpeis commented:
> is there a way to configure SPE to only collect taken branches?

What I believe you are asking here is to configure SPE to get us a pair of $\bf\textsf{\color{blue}SRC}$ -> $\textsf{\color{blue}TGT}$, where SRC and TGT are two taken branches. In other words, make SPE act as an LBR-like buffer with a branch stack depth of 1.

I don't think that is possible. SPE does some periodic capture of events packets, in our case branches.
Please consider the example below:
```
.  000007c0:  PC 0xAAA el2 ns=1
.  000007c9:  PAD
.  000007d3:  B COND
.  000007d5:  EV RETIRED NOT-TAKEN MISPRED
.  000007da:  LAT 12 ISSUE
.  000007dd:  LAT 13 TOT
.  000007e0:  TGT 0xBBB el2 ns=1
.  000007e9:  PAD
.  000007f7:  TS 12345
```

- `PC`: is the instruction that was captured, in our case the source branch
	- we can know whether the branch at `PC` was taken or not, and if it was a prediction miss/hit
	- this aligns with the [blogpost](https://community.arm.com/arm-community-blogs/b/tools-software-ides-blog/posts/introduction-to-statistical-profiling-support-in-streamline) you've provided
- `TGT`: is the target address of the landing block, ie where execution will continue next
	- actually this is not a branch
	- SPE has no information which the next branch will be, and whether it will be taken or not
	- we cannot configure SPE to do this
- `PBT`: optional HW feature pointing to the previous block  ([FEAT_SPE_PBT](https://developer.arm.com/documentation/109697/2024_12/Feature-descriptions/The-Armv8-7-architecture-extension))
	- it is a statistical profiling of the Previous Branch Target
	- TMU there is no known HW implementation of this optional feature.

I could re-word some points in the PR/patch to make the above more clear.

(@mikewilliams-arm feel free to correct me if I missed anything)

mikewilliams-arm commented:
> is there a way to configure SPE to only collect taken branches?

For the avoidance of doubt, and benefit of anyone finding this and reading it out of context, you *can* configure SPE to collect only taken branches, but only from FEAT_SPEv1p2. That's a relatively new feature in the field. From looking at the kernel sources, you need to check for `/sys/devices/arm_spe_0/format/inv_event_filter` and the syntax would be something like `perf record -e arm_spe_0/branch_filter=1,inv_event_filter=64/`. (I might be wrong - I don't have access to such a system.)

You can always do this filtering post-hoc in software. `perf record -e arm_spe_0/branch_filter=1/` should work on all SPE implementations, and according to Google's AI, 60% of branches are taken, so it's about a 66% overhead to store all the not taken branches and filter them out. `perf script --inject=b` used to do a poor job of preserving all the branch information through the injected events, making this harder to do. I believe that is something being looked into, if it's not already addressed.

However, even so, each sampled branch is exactly that - a single sampled branch. It does not collect sequences of branches other than through the aforementioned optional PBT extension. So, you can only infer that where you came from and where you branched to were executed.

paschalis-mpeis commented:
Great, thanks a lot Michael for filling in with details!

Indeed the differences are subtle. Just want to clarify that I've answered a slightly different question:
> What I believe you are asking here is to configure SPE to get us a pair of SRC -> TGT , where SRC and TGT are two taken branches. In other words, make SPE act as an LBR-like buffer with a branch stack depth of 1.

Whether we filter-out the non-taken branches at the HW collection level (with the `inv_event_filter` interface) or in post-processing SW, the information loss still holds:
> Please note that if we filter-out any non-taken branches at this point (ie `perf script`), then we'll exclude information we cannot later infer.

And regardless of filtering, we cannot end up with a pair of two taken branches, which could have left opportunities for inferring any FTs in between them.

ilinpv commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -1703,6 +1786,46 @@ std::error_code DataAggregator::parseBasicEvents() {
+      ++NumSpeBranchSamples;
+
+    registerSample(&SamplePair->first);
+    registerSample(&SamplePair->second);
```
Am I correct in understanding that it is the case when we have sample for branch SRC -> TGT which was or was not be taken. However we increase hotness of SRC and TGT nodes in any case registering samples always for both nodes and not taking into account ratio of samples with this branch taken and not taken?

paschalis-mpeis replied:
Hey Pavel,

Reading this back, you are concerned whether storing samples on TGT branches that are not NOT-TAKEN might increase hotness in a block that it shouldn't have. Correct?

That should not be a concern, as regardless of whether a branch is taken or not, the reported `TGT` is what was architecturally executed. In other words, `NOT-TAKEN` (or it's absence) characterizes what had happen in the src branch (`PC`), while `TGT` will always point to the path we end up taking.

So, for fall-through SPE packets, the `TGT` address would always be the next address from `PC` (ie, `0xA00` + `4`, which is the instruction size in AArch64):  
```
PC 0xA00
B COND
EV RETIRED NOT-TAKEN
TGT 0xA04
```

For taken branches, the `TGT` can be at a distance further than just `4` :  
```
PC 0xA00
B COND
EV RETIRED
TGT 0xBBB
```

In my previous examples I was using mock addresses for PC/TGT, so I've updated any relevant examples to avoid confusion.

ilinpv replied:
Right, thank you @paschalis-mpeis for clarifying about taken/not taken information and updating examples. @aaupov @maksfb would you like any additional explanations regarding SPE packets? Generally speaking SPE is providing event  based sampling for branches and doesn't have enough information to create trace of N>1 branches and inferring fall throughs. We are aiming to add BRBE ([Branch Record Buffer Extension](https://developer.arm.com/documentation/109697/2024_12/Feature-descriptions/The-Armv9-2-architecture-extension?lang=en )) support for this in BOLT and provide branch stack trace like LBR with it.

kaadam replied:
Hi, thanks Paschalis for your example. 
Maybe it's worth to highlight that the not-taken event is only related to conditional instruction (conditional branch or compare-and-branch), it tells that failed its condition code check, that's it. Since TGT (what you mentioned) "will always point to the path we end up taking", in this case presence of the not-taken event type is not relevant us, accordingly we will always get the 'taken paths'.  Theoretically these branch information support our optimization, bolt will be able to rely on them.

paschalis-mpeis replied:
Correct, thanks Adam. This is irrelevant to any unconditional branching (including call/ret).
Skipping 'non-taken' conditional branches is the optimization LBR/BRBE can do, as that can be inferred in post-processing.

paschalis-mpeis commented:
Just adding that we are in the process of upstreaming 'brstack' support for SPE, which would handle the `PBT` feature nicely for us, and would sit nicely within bolt sources. Once upstreamed, we can adapt the patch to work for the LBR-format.
(cc: @kaadam)

paschalis-mpeis commented:
Forced push to rebase to latest main to address conflicts (PCs/IPs were removed from the LBR samples).

Will be proceeding soon with an LBR patch, which for now will be stacked on top of this PR (cc: @kaadam).
